### PR TITLE
port GVCFWriter

### LINKED ISSUE #736: port GVCFWriter
needed for HaplotypeCaller

### PR DESCRIPTION
**Pull request by lbergelson**
Porting the GVCFWriter and tests

the first commit includes the original gatk classes modulo some import statements and package names
the next commit is my changes on top of it

closes #736

[Review on Reviewable](https://reviewable.io/reviews/broadinstitute/gatk/859)

### CODE REVIEW

**Comment by lbergelson** [comment]
@droazen I think you're the only one around to take this now.

**Comment by droazen** [comment]
Will review tomorrow, but first: why are we putting `GVCFWriter` here rather than in `htsjdk.variant` where it more naturally belongs?

**Comment by lbergelson** [comment]
Ah, good question.  Because adam opened a ticket for it..  It was originally separate from htsjdk because gvcf are a broad internal format and not intended for wider use.  Not sure if that's still really true.  We can always port to htsjdk in the future...

File: src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java [hunk 1]
```diff
@@ -662,4 +665,16 @@ public static double stddev(final double[] in, final int start, final int stop,
         for (int i = 0; i < is.length; ++i) ds[i] = is[i];
         return ds;
     }
+
+    /**
+     * Compute the median of a list of numbers
+     * @param values a list of numbers
+     * @return the median element
+     */
+    public static <T extends Number & Comparable<T>> double median(final Collection<T> values) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: as a heads up, the old version of HomRefBlock used this [implementation of median](https://github.com/broadinstitute/gsa-unstable/blob/master/public/gatk-utils/src/main/java/org/broadinstitute/gatk/utils/MathUtils.java#L975-L1002) which is not actually the median.  I've changed to the proper calculation of the median, which WILL produce different results.  It should hopefully be producing a better result, but it's also possible the other was intentional.

Comment (eitanbanks) [inline src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java]: Probably not intentional

**Comment by droazen** [comment]
Review complete -- back to @lbergelson

**Comment by akiezun** [comment]
@lbergelson what's the status here? is it ready to merge (modulo rebase)?

**Comment by lbergelson** [comment]
@droazen I've rebased this onto the current master.  Unfortunately, that seems to have totally blown away the comments that were on one of the existing commits...  

As best as I know I've addressed all of your comments.   I did a major rewrite on `HomRefBlock` since you last looked at it though, so you may have new comments.  

It also hasn't done a good job understanding that the files that moved from engine to utils are the same ones, so it's producing garbage diffs of them.  I can try to move them in the history if that would be helpful.

**Comment by lbergelson** [comment]
@akiezun Could you review this?  David never got a chance to.

**Comment by akiezun** [comment]
this looks good to me. But, having a *Writer class without any test that actually tests  file output is wrong. Please add one. back to @lbergelson

**Comment by akiezun** [comment]
other than that, good to merge

**Comment by akiezun** [comment]
@lbergelson I need this for HC. can you finish this?

**Comment by akiezun** [comment]
hmm @lbergelson, i think this requires a test that writes an actual GVCF file. It's a GVCFWriter after all

**Comment by lbergelson** [comment]
@akiezun I added one!

**Comment by lbergelson** [comment]
writeGVCFToDisk does writes it down and reads it back in and verifies that all the positions / GQ are correct

**Comment by lbergelson** [comment]
Also, you had comment about a test being removed, but the comment got lost before I replied to it.  The test that I removed was testing for a condition that got converted into an exception, and then avoided by restructuring the calling class to avoid having that condition come up.  I added tests to make sure that it throws properly in those cases though.

**Comment by akiezun** [comment]
we need a test that has an actual gvcf file and compares to a stored file.

**Comment by lbergelson** [comment]
I can add that, but it will only really tell us if something changes in the writer code, not that it's correct.  

Most of the logic that produces a gvcf is handled by haplotype caller.  The writer just process a stream of VariantContexts, merges adjacent homref calls into one block, and passes everything through to vcf writer.   

I write save some of these outputs and add a test that makes sure they don't change.  They look funny though since the VC being passed in aren't very realistic looking and only have the features that are examined by the writer.   Alternatively I can hand craft a more realistic looking file, but then we're just testing that I can hand write a gvcf.

**Comment by akiezun** [comment]
testing for no changes in the writer code is fine by me. 
can you take a small gvcf and hand-create it via the api? eg first few lines of https://github.com/broadinstitute/picard/pull/368/files#diff-81a922f111e9bbc567d31548dd9e481f

**Comment by lbergelson** [comment]
Yes.  Sounds reasonable.

**Pull request merged**